tasks/publish-to-cgw/publish-to-cgw.yaml

```diff
@@ -0,0 +1,175 @@
+---
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: publish-to-cgw
+  labels:
+    app.kubernetes.io/version: "0.1.0"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: release
+spec:
+  description: >-
+    Tekton task to publish content to CGW (Content-Gateway)
+  params:
+    - name: cgwHostname
+      type: string
+      description: >
+        The hostname of the content-gateway to publish the metadata to
+      default: https://developers.redhat.com/content-gateway/rest/admin
+    - name: cgwSecret
+      type: string
+      description: >
+        The kubernetes secret to use to authenticate to content-gateway.
+        It needs to contain two keys: username and token
+      default: publish-to-cgw-secret
+    - name: dataPath
+      type: string
+      description: >
+        Path to the JSON string of the merged data to use in the data workspace.
+    - name: contentDir
+      type: string
+      description: Path where the content to push is stored in the workspace
+    - name: subdirectory
+      description: Subdirectory inside the workspace to be used
+      type: string
+      default: ""
+  results:
+    - name: resultDataPath
+      type: string
+      description: |
+        The relative path in the workspace to the stored json of result data of the task
+  workspaces:
+    - name: data
+      description: Workspace to save the CR jsons to
+  steps:
+    - name: run-push-cgw-metadata
+      image: quay.io/konflux-ci/release-service-utils:38c3bfd7479c86b832cba5b61f9bbde40c469393
+      env:
+        - name: CGW_USERNAME
+          valueFrom:
+            secretKeyRef:
+              name: $(params.cgwSecret)
+              key: username
+        - name: CGW_TOKEN
+          valueFrom:
+            secretKeyRef:
+              name: $(params.cgwSecret)
+              key: token
+      script: |
+        #!/usr/bin/env bash
+        python3 <<EOF
+        """
+        Run `push-cgw-metadata` command to push metadata to CGW:
+        1. Extract all components under contentGateway key from dataPath
+        2. Find all the files in contentDir that starts with the component name
+        4. Generate necessary metadata for each file
+        5. Dump the metadata to a YAML file
+        6. Run `push-cgw-metadata` to push the metadata
+        """
+        import os
+        import json
+        import yaml
+        import hashlib
+        import subprocess
+
+        WORKSPACE_DIR="$(workspaces.data.path)/$(params.subdirectory)"
+        DATA_FILE = f"{WORKSPACE_DIR}/$(params.dataPath)"
```

Comment:
If you want to keep a task defined `params.subdirectory`, this isn't going to work. For example, the collect-data task is going to write the data file into like `$(workspaces.data.path)/foo/data.json`. The task result `dataPath` is going to be `foo/data.json`. So, the file is in `$(workspaces.data.path)/$(params.dataPath)` like here https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-advisory/create-advisory.yaml#L64

However, as defined here it is `$(workspaces.data.path)/$(params.subdirectory)/$(params.dataPath)` which doesn't exist unless subdirectory is `""` (I am not sure why have a default of `""` anyways). I would recommend either removing the subdirectory param and using dirname on the data file to get your proper subdir or at least changing this line to use `$(workspaces.data.path)` in place of `$WORKSPACE_DIR`

Reply:
I have made some updates as suggested.
`DATA_FILE = "$(workspaces.data.path)/$(params.dataPath)"`
and
`WORKSPACE_DIR = os.path.dirname(DATA_FILE)`

I have also removed `params.subdirectory` as we can now fetch that above. 

TY for the explanation, PTAL!

Reply:
> (I am not sure why have a default of "" anyways).

I had taken inspiration from some of the existing tasks, I assumed it will be always passed from the pipeline as `$(context.pipelineRun.uid)` 

In hindsight, I should not have set default at all if I expected it to be passed from the pipeline :)